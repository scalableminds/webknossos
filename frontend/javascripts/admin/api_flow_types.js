/**
 * api_flow_types.js
 * @flow
 */
import Enum from "Enumjs";

import type {
  BoundingBoxObject,
  Edge,
  CommentType,
  TreeGroup,
  RecommendedConfiguration,
} from "oxalis/store";
import type { ServerUpdateAction } from "oxalis/model/sagas/update_actions";
import type { SkeletonTracingStats } from "oxalis/model/accessors/skeletontracing_accessor";
import type { Vector3, Vector6, Point3 } from "oxalis/constants";

export type APIMessage = { ["info" | "warning" | "error"]: string };

export type ElementClass =
  | "uint8"
  | "uint16"
  | "uint24"
  | "uint32"
  | "uint64"
  | "float"
  | "double"
  | "int8"
  | "int16"
  | "int32"
  | "int64";

export type APIMapping = {
  +parent?: string,
  +name: string,
  +classes?: Array<Array<number>>,
  +colors?: Array<number>,
  +hideUnmappedIds?: boolean,
};

type APIDataLayerBase = {|
  +name: string,
  +boundingBox: BoundingBoxObject,
  +resolutions: Array<Vector3>,
  +elementClass: ElementClass,
|};

type APIColorLayer = {|
  ...APIDataLayerBase,
  +category: "color",
|};

export type APISegmentationLayer = {|
  ...APIDataLayerBase,
  +category: "segmentation",
  +largestSegmentId: number,
  +mappings?: Array<string>,
  +fallbackLayer?: ?string,
|};

export type APIDataLayer = APIColorLayer | APISegmentationLayer;

export type APIHistogramData = Array<{
  numberOfElements: number,
  elementCounts: Array<number>,
  min: number,
  max: number,
}>;

type APIDataSourceBase = {
  +id: {
    +name: string,
    +team: string,
  },
  +status?: string,
};

export type APIMaybeUnimportedDataSource = APIDataSourceBase & {
  +dataLayers?: Array<APIDataLayer>,
  +scale: ?Vector3,
};

export type APIDataSource = APIDataSourceBase & {
  +dataLayers: Array<APIDataLayer>,
  +scale: Vector3,
};

export type APIDataStore = {
  +name: string,
  +url: string,
  +isForeign: boolean,
  +isScratch: boolean,
  +isConnector: boolean,
};

export type APITracingStore = {
  +name: string,
  +url: string,
};

export type APITeam = {
  +id: string,
  +name: string,
  +organization: string,
};

type APIPublication = {
  +created: number,
  +description: string,
  +id: string,
  +imageUrl: string,
  +publicationDate: number,
  +title: string,
};

export type APIDatasetId = {
  +owningOrganization: string,
  +name: string,
};

export type APIDatasetDetails = {
  +species?: string,
  +brainRegion?: string,
  +acquisition?: string,
};

type APIDatasetBase = APIDatasetId & {
  +isUnreported: boolean,
  +allowedTeams: Array<APITeam>,
  +created: number,
  +dataStore: APIDataStore,
  +description: ?string,
  +details: ?APIDatasetDetails,
  +isEditable: boolean,
  +isPublic: boolean,
  +displayName: ?string,
  +logoUrl: ?string,
  +lastUsedByUser: number,
  +isForeign: boolean,
  +sortingKey: number,
  +publication: ?APIPublication,
};

export type APIMaybeUnimportedDataset = APIDatasetBase & {
  +dataSource: APIMaybeUnimportedDataSource,
  +isActive: boolean,
};

export type APIDataset = APIDatasetBase & {
  +dataSource: APIDataSource,
  +isActive: true,
};

export type APISampleDataset = {
  +name: string,
  +description: string,
  +status: "available" | "downloading" | "present",
};

export type APIDataSourceWithMessages = {
  +dataSource?: APIDataSource,
  +previousDataSource?: APIDataSource,
  +messages: Array<APIMessage>,
};

export type APITeamMembership = {
  +id: string,
  +name: string,
  +isTeamManager: boolean,
};

export type ExperienceMap = { +[string]: number };

export type ExperienceDomainList = Array<string>;

export type APIUserBase = {
  +email: string,
  +firstName: string,
  +lastName: string,
  +id: string,
  +isAnonymous: boolean,
  +teams: Array<APITeamMembership>,
};

export type APIUser = APIUserBase & {
  +created: number,
  +experiences: ExperienceMap,
  +isAdmin: boolean,
  +isActive: boolean,
  +isEditable: boolean,
  +lastActivity: number,
  +lastTaskTypeId: ?string,
  +organization: string,
};

export type APITimeInterval = {
  paymentInterval: {
    month: number,
    year: number,
  },
  durationInSeconds: number,
};
export type APIUserLoggedTime = {
  loggedTime: Array<APITimeInterval>,
};

export type APIActiveUser = {
  email: string,
  firstName: string,
  lastName: string,
  activeTasks: number,
};

export type APIRestrictions = {|
  +allowAccess: boolean,
  +allowUpdate: boolean,
  +allowFinish: boolean,
  +allowDownload: boolean,
  +allowSave?: boolean,
|};

export type APIAllowedMode = "orthogonal" | "oblique" | "flight" | "volume";

export type APISettings = {|
  +allowedModes: Array<APIAllowedMode>,
  +preferredMode?: APIAllowedMode,
  +branchPointsAllowed: boolean,
  +somaClickingAllowed: boolean,
  +mergerMode?: boolean,
  +allowedMagnifications?: {
    shouldRestrict: boolean,
    min?: number,
    max?: number,
  },
|};

export const APIAnnotationTypeEnum = Enum.make({
  Explorational: "Explorational",
  Task: "Task",
  View: "View",
  CompoundTask: "CompoundTask",
  CompoundProject: "CompoundProject",
  CompoundTaskType: "CompoundTaskType",
});

export type APIAnnotationType = $Keys<typeof APIAnnotationTypeEnum>;

export type TracingType = "skeleton" | "volume" | "hybrid";

export type APITaskType = {
  +id: string,
  +summary: string,
  +description: string,
  +teamId: string,
  +teamName: string,
  +settings: APISettings,
  +recommendedConfiguration: ?RecommendedConfiguration,
  +tracingType: TracingType,
};

export type TaskStatus = { +open: number, +active: number, +finished: number };

type APIScriptTypeBase = {
  +name: string,
  +gist: string,
};

export type APIScript = APIScriptTypeBase & {
  +id: string,
  +owner: APIUserBase,
};

export type APIScriptUpdater = APIScriptTypeBase & {
  +id: string,
  +owner: string,
};

export type APIScriptCreator = APIScriptTypeBase & {
  +owner: string,
};

type APIProjectTypeBase = {
  +name: string,
  +team: string,
  +priority: number,
  +paused: boolean,
  +expectedTime: number, // Also known as "time limit"
  +isBlacklistedFromReport: boolean,
};

export type APIProject = APIProjectTypeBase & {
  +id: string,
  +owner: APIUserBase,
};

export type APIProjectUpdater = APIProjectTypeBase & {
  +id: string,
  +owner: string,
};

export type APIProjectCreator = APIProjectTypeBase & {
  +owner: string,
};

export type APIProjectWithAssignments = APIProject & {
  +numberOfOpenAssignments: number,
};

export type APITask = {
  +boundingBox: ?BoundingBoxObject,
  +boundingBoxVec6?: Vector6,
  +created: number,
  +creationInfo: ?string,
  +dataSet: string,
  +editPosition: Vector3,
  +editRotation: Vector3,
  +formattedHash: string,
  +id: string,
  +neededExperience: {
    +domain: string,
    +value: number,
  },
  +projectName: string,
  +script: ?APIScript,
  +status: TaskStatus,
  +team: string,
  +tracingTime: ?number,
  +type: APITaskType,
  +directLinks?: Array<string>,
};

export type APIAnnotationCompact = {
  +tracing: {
    +skeleton: ?string,
    +volume: ?string,
  },
  +dataSetName: string,
  +organization: string,
  +description: string,
  +formattedHash: string,
  +modified: number,
  +id: string,
  +isPublic: boolean,
  +name: string,
  +state: string,
  +stats: SkeletonTracingStats | {||},
  +tags: Array<string>,
  +tracingTime: ?number,
  +typ: APIAnnotationType,
};

export type LocalMeshMetaData = {|
  isVisible?: boolean,
  isLoaded?: boolean,
  isLoading?: boolean,
|};

export type RemoteMeshMetaData = {|
  annotationId: string,
  position: Vector3,
  description: string,
  id: string,
|};

export type MeshMetaData = {|
  ...LocalMeshMetaData,
  ...RemoteMeshMetaData,
|};

type APIAnnotationBase = APIAnnotationCompact & {
  +dataStore: APIDataStore,
  +tracingStore: APITracingStore,
  +restrictions: APIRestrictions,
  +settings: APISettings,
  +user?: APIUserBase,
  +meshes: Array<MeshMetaData>,
};

export type APIAnnotation = APIAnnotationBase & {
  +task: ?APITask,
};

export type APIAnnotationWithTask = APIAnnotationBase & {
  +task: APITask,
};

export type APITaskWithAnnotation = APITask & {
  +annotation: APIAnnotation,
};

export type DatasetConfig = {
  +name: string,
  +organization: string,
  +datastore: string,
  +zipFile: File,
};

type NeuroglancerLayer = {
  // This is the source URL of the layer, should start with gs://, http:// or https://
  source: string,
  type: "image" | "segmentation",
};

type NeuroglancerDatasetConfig = {
  [organizationName: string]: {
    [datasetName: string]: {
      layers: { [layerName: string]: NeuroglancerLayer },
      credentials?: Object,
    },
  },
};

type BossDatasetConfig = {
  [organizationName: string]: {
    [datasetName: string]: {
      domain: string,
      collection: string,
      experiment: string,
      username: string,
      password: string,
    },
  },
};

export type WkConnectDatasetConfig = {
  neuroglancer?: NeuroglancerDatasetConfig,
  boss?: BossDatasetConfig,
};

export type APITimeTracking = {
  time: string,
  timestamp: number,
  annotation: string,
  _id: string,
  task_id: string,
  project_name: string,
  tasktype_id: string,
  tasktype_summary: string,
};

export type APIProjectProgressReport = {
  +projectName: string,
  +paused: boolean,
  +totalTasks: number,
  +totalInstances: number,
  +openInstances: number,
  +activeInstances: number,
  +finishedInstances: number,
  +priority: number,
  +billedMilliseconds: number,
};

export type APIOpenTasksReport = {
  +id: string,
  +user: string,
  +totalAssignments: number,
  +assignmentsByProjects: { [projectName: string]: number },
};

export type APIOrganization = {
  +id: string,
  +name: string,
  +additionalInformation: string,
  +displayName: string,
  +enableAutoVerify: boolean,
};

export type APIBuildInfo = {
  webknossos: {
    name: string,
    commitHash: string,
    scalaVersion: string,
    version: string,
    sbtVersion: string,
    commitDate: string,
    ciTag: string,
    ciBuild: string,
    gitTag: string,
    version: string,
    datastoreApiVersion: string,
  },
  "webknossos-wrap": {
    builtAtMillis: string,
    name: string,
    commitHash: string,
    scalaVersion: string,
    version: string,
    sbtVersion: string,
    builtAtString: string,
  },
  webknossosDatastore?: {
    name: string,
    commitHash: string,
    scalaVersion: string,
    version: string,
    sbtVersion: string,
    commitDate: string,
    ciTag: string,
    ciBuild: string,
    gitTag: string,
    version: string,
    datastoreApiVersion: string,
  },
};

export type APIFeatureToggles = {
  +discussionBoard: string | false,
  +discussionBoardRequiresAdmin: boolean,
  +defaultOrganization: string,
  +addForeignDataset: boolean,
  +hideNavbarLogin: boolean,
  +isDemoInstance: boolean,
  +autoBrushReadyDatasets: Array<string>,
  +isDemoInstance: boolean,
};

// Tracing related datatypes
export type APIUpdateActionBatch = {
  version: number,
  value: Array<ServerUpdateAction>,
};

export type ServerNode = {
  id: number,
  position: Point3,
  rotation: Point3,
  bitDepth: number,
  viewport: number,
  resolution: number,
  radius: number,
  createdTimestamp: number,
  interpolation: boolean,
};

export type ServerBranchPoint = {
  createdTimestamp: number,
  nodeId: number,
};

export type ServerBoundingBox = {
  topLeft: Point3,
  width: number,
  height: number,
  depth: number,
};

export type ServerBoundingBoxTypeTuple = {
  topLeft: Vector3,
  width: number,
  height: number,
  depth: number,
};

export type ServerSkeletonTracingTree = {
  branchPoints: Array<ServerBranchPoint>,
  color: ?{ r: number, g: number, b: number },
  comments: Array<CommentType>,
  edges: Array<Edge>,
  name: string,
  nodes: Array<ServerNode>,
  treeId: number,
  createdTimestamp: number,
  groupId?: ?number,
  isVisible?: boolean,
};

export type ServerTracingBase = {|
  id: string,
  userBoundingBox?: ServerBoundingBox,
  createdTimestamp: number,
  dataSetName: string,
  editPosition: Point3,
  editRotation: Point3,
  error?: string,
  version: number,
  zoomLevel: number,
|};

export type ServerSkeletonTracing = {|
  ...ServerTracingBase,
  activeNodeId?: number,
  boundingBox?: ServerBoundingBox,
  trees: Array<ServerSkeletonTracingTree>,
  treeGroups: ?Array<TreeGroup>,
|};

export type ServerVolumeTracing = {|
  ...ServerTracingBase,
  activeSegmentId?: number,
  boundingBox: ServerBoundingBox,
  elementClass: ElementClass,
  fallbackLayer?: string,
  largestSegmentId: number,
|};

export type ServerTracing = ServerSkeletonTracing | ServerVolumeTracing;

export type HybridServerTracing = {
  skeleton: ?ServerSkeletonTracing,
  volume: ?ServerVolumeTracing,
};

export default {};
