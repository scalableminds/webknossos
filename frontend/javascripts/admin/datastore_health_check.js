// @flow
import _ from "lodash";

import Request from "libs/request";
import * as RestAPI from "admin/admin_rest_api";
import Toast from "libs/toast";
import messages from "messages";

// Create a throttled function which depends on its arguments.
// That way, each datastore is checked for health in a throttled and isolated manner
const memoizedThrottle = (func, wait = 0, options = {}): Function => {
  // Memoize the creation of a throttling function
  const mem = _.memoize(() => _.throttle(func, wait, options), options.resolver);
  return (...args: Array<*>) => {
    // look up (or create) the throttling function and invoke it
    mem(...args)(...args);
  };
};

// Do not call this function directly, but call pingMentionedDataStores instead
// which will take care of extracting the hostnames.
// Otherwise, the memoization will not work correctly if the path and query-string are part of the URL
const pingDataStoreIfAppropriate = memoizedThrottle(async (requestedUrl: string): Promise<*> => {
  const [datastores, tracingstore, isInMaintenance] = await Promise.all([
    RestAPI.getDataStoresCached(),
    RestAPI.getTracingStoreCached(),
    RestAPI.isInMaintenance(),
  ]).catch(() => [null, null, null]);
  if (datastores == null || tracingstore == null || isInMaintenance == null) {
    Toast.warning(messages.offline);
    return;
  }

  const stores = datastores
    .map(datastore => ({ ...datastore, path: "data" }))
    .concat({ ...tracingstore, path: "tracings" });
  if (isInMaintenance) {
    Toast.warning(messages.planned_maintenance);
  } else {
    const usedStore = stores.find(ds => requestedUrl.indexOf(ds.url) > -1);
    if (usedStore != null) {
      const { url, path } = usedStore;
      const healthEndpoint = `${url}/${path}/health`;
      Request.triggerRequest(healthEndpoint, {
        doNotInvestigate: true,
        mode: "cors",
        timeout: 5000,
      }).then(
        () => checkVersionMismatch(url),
        () => Toast.warning(messages["datastore.health"]({ url })),
      );
    }
  }
}, 5000);

async function checkVersionMismatch(url: string) {
  const [buildinfoWebknossos, buildinfoDatastore] = await Promise.all([
    RestAPI.getBuildInfo(),
    RestAPI.getDataStoreBuildInfo(url),
  ]);

  const expectedDatastoreApiVersion = buildinfoWebknossos.webknossos.datastoreApiVersion;

  const buildInfoWebknossosDatastore = buildinfoDatastore.webknossosDatastore
    ? buildinfoDatastore.webknossosDatastore
    : buildinfoDatastore.webknossos;
  const suppliedDatastoreApiVersion = buildInfoWebknossosDatastore.datastoreApiVersion;

  if (
    Number(expectedDatastoreApiVersion.split(".")[0]) <
    Number(suppliedDatastoreApiVersion.split(".")[0])
  ) {
    Toast.warning(
      messages["datastore.version.too_new"]({
        expectedDatastoreApiVersion,
        suppliedDatastoreApiVersion,
        url,
      }),
    );
  } else if (
    Number(expectedDatastoreApiVersion.split(".")[0]) >
    Number(suppliedDatastoreApiVersion.split(".")[0])
  ) {
    Toast.warning(
      messages["datastore.version.too_old"]({
        expectedDatastoreApiVersion,
        suppliedDatastoreApiVersion,
        url,
      }),
    );
  }
}

const extractUrls = (str: string): Array<string> => {
  const urlMatcher = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{2,256}(\.[a-z]{2,6})?\b([-a-zA-Z0-9@:%_+.~#?&\\=]*)/g;
  return _.uniq(str.match(urlMatcher) || []);
};

export const pingMentionedDataStores = (str: string): void => {
  extractUrls(str).map(pingDataStoreIfAppropriate);
};

export default {};
