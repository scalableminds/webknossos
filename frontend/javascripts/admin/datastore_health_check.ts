import * as RestAPI from "admin/rest_api";
import Toast from "libs/toast";
import _ from "lodash";
import messages from "messages";
import type { APIBuildInfoDatastore, APIBuildInfoWk } from "types/api_types";

// Create a throttled function which depends on its arguments.
// That way, each datastore is checked for health in a throttled and isolated manner
const memoizedThrottle = <F extends (...args: Array<any>) => any>(func: F, wait = 0): F => {
  // Memoize the creation of a throttling function
  const mem = _.memoize((..._args: any[]) => _.throttle(func, wait));

  return ((...args: Parameters<F>) => {
    // look up (or create) the throttling function and invoke it
    return mem(...args)(...args);
  }) as F;
};

// Do not call this function directly, but call pingMentionedDataStores instead
// which will take care of extracting the hostnames.
// Otherwise, the memoization will not work correctly if the path and query-string are part of the URL
const pingDataStoreIfAppropriate = memoizedThrottle(async (requestedUrl: string): Promise<any> => {
  const [datastores, tracingstore, isInMaintenance] = await Promise.all([
    RestAPI.getDataStoresCached(),
    RestAPI.getTracingStoreCached(),
    RestAPI.isInMaintenance(),
  ]).catch(() => [null, null, null]);

  if (datastores == null || tracingstore == null || isInMaintenance == null) {
    Toast.warning(messages.offline);
    return;
  }

  const stores: Array<{ url: string; path: "tracings" | "data" }> = [
    { ...tracingstore, path: "tracings" },
    ...datastores.map((datastore) => ({ ...datastore, path: "data" as const })),
  ] as const;

  if (isInMaintenance) {
    Toast.warning(messages.planned_maintenance);
  } else {
    const usedStore = stores.find((ds) => requestedUrl.indexOf(ds.url) > -1);

    if (usedStore != null) {
      const { url, path } = usedStore;
      RestAPI.pingHealthEndpoint(url, path).then(
        () => {
          if (usedStore.path === "data") {
            checkVersionMismatchInDataStore(url);
          }
        },
        () =>
          Toast.warning(
            messages["datastore.health"]({
              url,
            }),
          ),
      );
    }
  }
}, 5000);

async function checkVersionMismatchInDataStore(datastoreUrl: string) {
  const [buildinfoWebknossos, buildinfoDatastore] = (await Promise.all([
    RestAPI.getBuildInfo(),
    RestAPI.getDataOrTracingStoreBuildInfo(datastoreUrl),
  ])) as [APIBuildInfoWk, APIBuildInfoDatastore];
  const expectedDatastoreApiVersion = buildinfoWebknossos.webknossos.datastoreApiVersion;
  const buildInfoWebknossosDatastore = buildinfoDatastore.webknossosDatastore;
  const suppliedDatastoreApiVersion = buildInfoWebknossosDatastore.datastoreApiVersion;

  if (
    Number(expectedDatastoreApiVersion.split(".")[0]) <
    Number(suppliedDatastoreApiVersion.split(".")[0])
  ) {
    Toast.warning(
      messages["datastore.version.too_new"]({
        expectedDatastoreApiVersion,
        suppliedDatastoreApiVersion,
        datastoreUrl,
      }),
    );
  } else if (
    Number(expectedDatastoreApiVersion.split(".")[0]) >
    Number(suppliedDatastoreApiVersion.split(".")[0])
  ) {
    Toast.warning(
      messages["datastore.version.too_old"]({
        expectedDatastoreApiVersion,
        suppliedDatastoreApiVersion,
        datastoreUrl,
      }),
    );
  }
}

const extractUrls = (str: string): Array<string> => {
  const urlMatcher =
    /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._+~#=]{2,256}(\.[a-z]{2,6})?\b([-a-zA-Z0-9@:%_+.~#?&\\=]*)/g;
  return _.uniq(str.match(urlMatcher) || []);
};

export const pingMentionedDataStores = (str: string): void => {
  extractUrls(str).map(pingDataStoreIfAppropriate);
};
export default {};
