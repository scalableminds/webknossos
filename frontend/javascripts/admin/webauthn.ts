import * as webauthn from "@github/webauthn-json";
import { getActiveUser, getOrganization } from "admin/rest_api";
import Request from "libs/request";
import type { APIOrganization, APIUser } from "types/api_types";

export type WebAuthnKeyDescriptor = {
  id: string;
  name: string;
};

export async function doWebAuthnLogin(): Promise<[APIUser, APIOrganization]> {
  const webAuthnAuthAssertion = await Request.receiveJSON("/api/auth/webauthn/auth/start", {
    method: "POST",
  });
  const response = await webauthn.get({
    publicKey: webAuthnAuthAssertion,
  });
  await Request.sendJSONReceiveJSON("/api/auth/webauthn/auth/finalize", {
    method: "POST",
    data: { key: response },
  });

  const activeUser = await getActiveUser();
  const organization = await getOrganization(activeUser.organization);
  return [activeUser, organization];
}

export async function doWebAuthnRegistration(name: string): Promise<any> {
  const webAuthnRegistrationAssertion = await Request.receiveJSON(
    "/api/auth/webauthn/register/start",
    {
      method: "POST",
    },
  );

  const credential = await webauthn.create({
    publicKey: webAuthnRegistrationAssertion,
  });

  return Request.sendJSONReceiveJSON("/api/auth/webauthn/register/finalize", {
    data: {
      name: name,
      key: credential,
    },
    method: "POST",
  });
}

export async function listWebAuthnKeys(): Promise<Array<WebAuthnKeyDescriptor>> {
  return await Request.receiveJSON("/api/auth/webauthn/keys");
}

export async function removeWebAuthnKey(key: WebAuthnKeyDescriptor): Promise<void> {
  await Request.sendJSONReceiveArraybuffer("/api/auth/webauthn/keys", {
    method: "DELETE",
    data: key,
  });
}
