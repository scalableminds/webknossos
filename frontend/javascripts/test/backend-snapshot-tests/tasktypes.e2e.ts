import {
  tokenUserA,
  setCurrToken,
  resetDatabase,
  writeTypeCheckingFile,
} from "test/enzyme/e2e-setup";
import * as api from "admin/admin_rest_api";
import test from "ava";
import { APIAllowedMode } from "types/api_flow_types";
test.before("Change token", async () => {
  resetDatabase();
  setCurrToken(tokenUserA);
});
test("getTaskTypes()", async (t) => {
  const taskTypes = await api.getTaskTypes();
  writeTypeCheckingFile(taskTypes, "task-type", "APITaskType", {
    isArray: true,
  });
  t.snapshot(taskTypes, {
    id: "taskTypes-getTaskTypes",
  });
});
test("getTaskType()", async (t) => {
  const taskTypes = await api.getTaskTypes();
  const taskType = await api.getTaskType(taskTypes[0].id);
  t.snapshot(taskType, {
    id: "taskTypes-getTaskType",
  });
});
test("createTaskType and deleteTaskType", async (t) => {
  const activeUser = await api.getActiveUser();
  const aTeam = activeUser.teams[0];
  const newTaskType = {
    id: null,
    summary: "summary",
    description: "description",
    teamId: aTeam.id,
    settings: {
      somaClickingAllowed: true,
      branchPointsAllowed: true,
      mergerMode: false,
      volumeInterpolationAllowed: false,
      allowedModes: ["orthogonal" as APIAllowedMode],
      resolutionRestrictions: {},
    },
    recommendedConfiguration: null,
    tracingType: "skeleton" as const,
  };
  const createdTaskType = await api.createTaskType(newTaskType);
  // Since the id will change after re-runs, we fix it here for easy
  // snapshotting
  const createdTaskTypeWithFixedId = Object.assign({}, createdTaskType, {
    id: "fixed-id",
  });
  t.snapshot(createdTaskTypeWithFixedId, {
    id: "taskTypes-createdTaskType",
  });
  const response = await api.deleteTaskType(createdTaskType.id);
  t.snapshot(response, {
    id: "taskTypes-deleteTaskType",
  });
});
test("updateTaskType()", async (t) => {
  const taskTypes = await api.getTaskTypes();
  const taskType = await api.getTaskType(taskTypes[0].id);
  const updatedTaskType = await api.updateTaskType(
    taskType.id,
    Object.assign({}, taskType, {
      summary: "new-test-summary",
    }),
  );
  t.snapshot(updatedTaskType, {
    id: "taskTypes-updateTaskType",
  });
  // Change back
  const revertedTaskType = await api.updateTaskType(taskType.id, taskType);
  t.snapshot(revertedTaskType, {
    id: "taskTypes-revertedTaskType",
  });
});
