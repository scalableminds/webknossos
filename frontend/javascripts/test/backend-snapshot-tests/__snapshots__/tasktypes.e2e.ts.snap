// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Task Types API (E2E) > createTaskType and deleteTaskType 1`] = `
{
  "description": "description",
  "id": "fixed-id",
  "recommendedConfiguration": null,
  "settings": {
    "allowedModes": [
      "orthogonal",
    ],
    "branchPointsAllowed": true,
    "magRestrictions": {},
    "mergerMode": false,
    "somaClickingAllowed": true,
    "volumeInterpolationAllowed": false,
  },
  "summary": "summary",
  "teamId": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
  "tracingType": "skeleton",
}
`;

exports[`Task Types API (E2E) > createTaskType and deleteTaskType 2`] = `
{
  "messages": [
    {
      "success": "Task type “summary” successfully deleted",
    },
  ],
}
`;

exports[`Task Types API (E2E) > getTaskType() 1`] = `
{
  "description": "Check those cells out!",
  "id": "570b9f4c2a7c0e4c008da6ee",
  "recommendedConfiguration": null,
  "settings": {
    "allowedModes": [
      "orthogonal",
      "oblique",
      "flight",
    ],
    "branchPointsAllowed": true,
    "magRestrictions": {},
    "mergerMode": false,
    "somaClickingAllowed": true,
    "volumeInterpolationAllowed": false,
  },
  "summary": "ek_0563_BipolarCells",
  "teamId": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
  "tracingType": "skeleton",
}
`;

exports[`Task Types API (E2E) > getTaskTypes() 1`] = `
[
  {
    "description": "Check those cells out!",
    "id": "570b9f4c2a7c0e4c008da6ee",
    "recommendedConfiguration": null,
    "settings": {
      "allowedModes": [
        "orthogonal",
        "oblique",
        "flight",
      ],
      "branchPointsAllowed": true,
      "magRestrictions": {},
      "mergerMode": false,
      "somaClickingAllowed": true,
      "volumeInterpolationAllowed": false,
    },
    "summary": "ek_0563_BipolarCells",
    "teamId": "570b9f4b2a7c0e3b008da6ec",
    "teamName": "team_X1",
    "tracingType": "skeleton",
  },
  {
    "description": "Check those cells out!",
    "id": "570b9f4c2a7c0e4c008da6ff",
    "recommendedConfiguration": null,
    "settings": {
      "allowedModes": [
        "orthogonal",
        "oblique",
        "flight",
      ],
      "branchPointsAllowed": true,
      "magRestrictions": {},
      "mergerMode": true,
      "somaClickingAllowed": true,
      "volumeInterpolationAllowed": false,
    },
    "summary": "ek_0674_BipolarCells",
    "teamId": "69882b370d889b84020efd4f",
    "teamName": "team_X2",
    "tracingType": "skeleton",
  },
]
`;

exports[`Task Types API (E2E) > updateTaskType() 1`] = `
{
  "description": "Check those cells out!",
  "id": "570b9f4c2a7c0e4c008da6ee",
  "messages": [
    {
      "success": "Task type successfully edited",
    },
  ],
  "recommendedConfiguration": null,
  "settings": {
    "allowedModes": [
      "orthogonal",
      "oblique",
      "flight",
    ],
    "branchPointsAllowed": true,
    "magRestrictions": {},
    "mergerMode": false,
    "somaClickingAllowed": true,
    "volumeInterpolationAllowed": false,
  },
  "summary": "new-test-summary",
  "teamId": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
  "tracingType": "skeleton",
}
`;

exports[`Task Types API (E2E) > updateTaskType() 2`] = `
{
  "description": "Check those cells out!",
  "id": "570b9f4c2a7c0e4c008da6ee",
  "messages": [
    {
      "success": "Task type successfully edited",
    },
  ],
  "recommendedConfiguration": null,
  "settings": {
    "allowedModes": [
      "orthogonal",
      "oblique",
      "flight",
    ],
    "branchPointsAllowed": true,
    "magRestrictions": {},
    "mergerMode": false,
    "somaClickingAllowed": true,
    "volumeInterpolationAllowed": false,
  },
  "summary": "ek_0563_BipolarCells",
  "teamId": "570b9f4b2a7c0e3b008da6ec",
  "teamName": "team_X1",
  "tracingType": "skeleton",
}
`;
